test/integration/src/test/java/io/pravega/test/integration/WatermarkingTest.java

```diff
@@ -0,0 +1,244 @@
+        @Cleanup
+        EventStreamWriter<Long> writer1 = clientFactory.createEventWriter(stream, javaSerializer,
+                EventWriterConfig.builder().build());
+        @Cleanup
+        EventStreamWriter<Long> writer2 = clientFactory.createEventWriter(stream, javaSerializer,
+                EventWriterConfig.builder().build());
+
+        AtomicBoolean stopFlag = new AtomicBoolean(false);
+        // write events
+        writeEvents(writer1, stopFlag);
+        writeEvents(writer2, stopFlag);
+
+        // scale the stream several times so that we get complex positions
+        scale(controller, streamObj, config);
+
+        @Cleanup
+        ConnectionFactory connectionFactory = new ConnectionFactoryImpl(clientConfig);
+        @Cleanup
+        SynchronizerClientFactory syncClientFactory = SynchronizerClientFactory.withScope(scope, clientConfig);
+
+        String markStream = NameUtils.getMarkStreamForStream(stream);
+        RevisionedStreamClient<Watermark> watermarkReader = syncClientFactory.createRevisionedStreamClient(markStream,
+                new WatermarkSerializer(),
+                SynchronizerConfig.builder().build());
+        AssertExtensions.assertEventuallyEquals(true, () -> {
+            Iterator<Map.Entry<Revision, Watermark>> watermarks = watermarkReader.readFrom(watermarkReader.fetchOldestRevision());
+            return watermarks.hasNext();
+        }, 30000);
+        Iterator<Map.Entry<Revision, Watermark>> watermarks = watermarkReader.readFrom(watermarkReader.fetchOldestRevision());
+        Watermark watermark = watermarks.next().getValue();
+        
+        stopFlag.set(true);
+
+        // read events from the stream
+        @Cleanup
+        ReaderGroupManager readerGroupManager = new ReaderGroupManagerImpl(scope, controller, syncClientFactory, connectionFactory);
+        String readerGroup = "rg";
+
+        long timeLow = watermark.getLowerTimeBound();
+        Map<Segment, Long> positionMap0 = watermark.getStreamCut()
+                                                   .entrySet().stream().collect(Collectors.toMap(x -> new Segment(scope, stream, x.getKey().getSegmentId()),
+                        Map.Entry::getValue));
+
+        StreamCut streamCutStart = new StreamCutImpl(streamObj, positionMap0);
+        Map<Stream, StreamCut> start = Collections.singletonMap(streamObj, streamCutStart);
+        
+        readerGroupManager.createReaderGroup(readerGroup, ReaderGroupConfig.builder().stream(streamObj)
+                                                                           .startingStreamCuts(start).build());
+
+        // create reader on the stream
+        @Cleanup
+        final EventStreamReader<Long> reader = clientFactory.createReader("myreader",
+                readerGroup,
+                javaSerializer,
+                ReaderConfig.builder().build());
+        
+        // read events from the reader. 
```

Comment:
It would be good to assert that the reader receives a timeWindow from the `getTimeWindow` that has and that after all the events are read the low watermark is the largest of the low watermarks in the mark segment.

Reply:
It once it is passed the noted time it should eventually return a watermark passed it. This can be asserted using assertEventuallyEquals with a readNextEvent() call in there to give it a chance to update the group state.